### PR TITLE
Add `array::IntoIter::{empty, from_raw_parts}`

### PR DESCRIPTION
**Pull request by scottmcm**
`array::IntoIter` has a bunch of really handy logic for dealing with partial arrays, but it's currently hamstrung by only being creatable from a fully-initialized array.

This PR adds two new constructors:
- a safe & const `empty`, since `[].into_iter()` can only give `IntoIter<T, 0>`, not `IntoIter<T, N>`.
- an unsafe `from_raw_parts`, to allow experimentation with new uses.

(Slice & vec iterators don't need `from_raw_parts` because you `from_raw_parts` the slice or vec instead, but there's no useful way to made a `<[T; N]>::from_raw_parts`, so I think this is a reasonable place to have one.)

### CODE REVIEW

File: library/core/src/array/iter.rs [hunk 1]
```diff
@@ -82,6 +82,135 @@ impl<T, const N: usize> IntoIter<T, N> {
...
+    /// let empty = IntoIter::<std::convert::Infallible, 200>::empty();
+    /// assert_eq!(empty.len(), 0);
+    /// ```
+    ///
+    /// `[1, 2].into_iter()` and `[].into_iter()` have different types
+    /// ```should_fail
+    /// #![feature(array_into_iter_constructors)]
+    /// use std::array::IntoIter;
+    ///
+    /// # // FIXME: use `.into_iter()` once the doc tests are in edition2021
```
Comment (jyn514) [inline library/core/src/array/iter.rs:178]: You can add `edition2021` at the top (next to should_fail).

Comment (scottmcm) [reply library/core/src/array/iter.rs]: Oh, right!  Thank you.

File: library/core/src/array/iter.rs [hunk 2]
```diff
@@ -82,6 +82,134 @@ impl<T, const N: usize> IntoIter<T, N> {
...
+    /// }
+    ///
+    /// let r: [_; 4] = next_chunk(&mut (10..16)).unwrap();
+    /// assert_eq!(r, [10, 11, 12, 13]);
+    /// let r: IntoIter<_, 40> = next_chunk(&mut (10..16)).unwrap_err();
+    /// assert_eq!(r.collect::<Vec<_>>(), vec![10, 11, 12, 13, 14, 15]);
+    /// ```
+    #[unstable(feature = "array_into_iter_constructors", issue = "88888888")]
+    #[rustc_const_unstable(feature = "array_into_iter_constructors_const", issue = "88888888")]
+    pub const unsafe fn from_raw_parts(
```
Comment (kennytm) [inline library/core/src/array/iter.rs:144]: calling the function "from_<b>raw</b>_parts" sounds strange, as "raw parts" is mostly used for composing pointers into slices.

i think it is better to call it `from_parts`?

File: library/core/src/array/iter.rs [hunk 3]
```diff
@@ -82,6 +82,134 @@ impl<T, const N: usize> IntoIter<T, N> {
...
+    ///     } else {
+    ///         IntoIter::empty()
+    ///     }
+    /// }
+    ///
+    /// assert_eq!(get_bytes(true).collect::<Vec<_>>(), vec![1, 2, 3, 4]);
+    /// assert_eq!(get_bytes(false).collect::<Vec<_>>(), vec![]);
+    /// ```
+    #[unstable(feature = "array_into_iter_constructors", issue = "88888888")]
+    pub fn empty() -> Self {
```
Comment (kennytm) [inline library/core/src/array/iter.rs:204]: can this be `const` 🤔. is it blocked by using generics in const fn?

**Review by kennytm — COMMENTED** [review]
lgtm but the issue number should be changed from 88888888

File: library/core/src/array/iter.rs (hunk 2 above)
Comment (scottmcm) [reply library/core/src/array/iter.rs:144]: Hmm, I was thinking of the `Vec` and `String` ones, which aren't *exactly* slices (though they're certainly highly related).

I guess the other direction would be something like `new_unchecked` or `from_buffer`?  Do we have any existing methods that are a good precedent?

Comment (kennytm) [reply library/core/src/array/iter.rs:146]: `new_unchecked` sounds good. (there can be further discussion before stabilization for sure)

**Pull request merged**